include/sempr/entity/RDFEntity.hpp

```diff
@@ -1,43 +1,79 @@
-#ifndef SEMPR_ENTITY_RDFENTITY_H_
-#define SEMPR_ENTITY_RDFENTITY_H_
+#ifndef SEMPR_ENTITY_RDFENTITY_HPP_
+#define SEMPR_ENTITY_RDFENTITY_HPP_
 
-#include <odb/core.hxx>
 #include <sempr/core/RDF.hpp>
 #include <sempr/entity/Entity.hpp>
 #include <sempr/entity/Triple.hpp>
 
-// #include <sempr/core/EventBroker.hpp>
 #include <sempr/core/EntityEvent.hpp>
 
-#include <vector>
-
 namespace sempr { namespace entity {
 
+
+class TripleIterator_impl {
+    friend class TripleIterator;
+protected:
+    virtual ~TripleIterator_impl();
```

Comment:
The problem is that I want to return an iterator by value, and not a pointer to an iterator that points to some triples. The latter would force the user to write stuff like
`(**it).subject;` or `*(*it)->subject` and to delete the iterator in the end. If I'd just return a FooIterator, but the method signature in the RDFEntity class specifies a TripleIterator, the result would be a TripleIterator created from a FooIterator... not what I wanted. So my solution is to give the TripleIterator class a pointer to the actual implementation which has to be provided by the entity types that are RDF-iterable. The `TripleIterator_impl` is the interface for this implementation, which we need for the TripleIterator to work correctly (virtual methods...).

I agree that the naming of the actual implementations `[...]_impl` is a bit weird, but it simply follows the pattern. Do you have any suggestion how to improve this?

Reply:
In this case I think the `TripleIterator` shall be renamed because its an abstract wrapper for the real iterators. It could be renamed to something like `AbstractTripleIterator` or `TripleIteratorWrapper` after these step the `*_impl` could be removed.

Reply:
Agreed. I'm on it.